### ISSUE TITLE
Closing server while client is sending data can result in runaway thread and memory leak

### ISSUE THREAD

**Issue by mcmellawatt** [Azure/amqpnetlite#87]
I have discovered a rare timing issue that results in a runaway thread and memory leak.  The issue occurs rarely in production, but I have built a test client and test server which reliably reproduces it on my system.

The problem arises when the client is in the middle of a write and the server closes its ContainerHost.

The series of events are as follows, marked T1 and T2 for thread 1 (writing thread) and thread 2 (pump thread).

(T1) 1. The client calls SenderLink.Send() which sets the SenderLink.writing boolean to true
(T2) 2. The server closes its ContainerHost which triggers abort logic in the client, stemming from the Connection.PumpThread() thread which calls Connection.OnFrame()
(T2) 3. This abort logic sets the Connection.state to State.End in the client
(T1) 4. The client calls SenderLink.Session.SendDelivery() in the SenderLink.WriteDelivery() method, which throws an AmqpException because the connection is in an end state
(T2) 5. Because an exception is thrown in step 4, the writing boolean in the SenderLink is never set back to false, and the abort logic in the pump thread eventually calls SenderLink.OnAbort() which spins forever because writing is set to true.

As far as I can tell, there is no way to terminate this pump thread now that it is in this state i.e. spinning in SenderLink.OnAbort()

Again, the problem is a rare timing issue and requires the connection state to be set to State.End after SenderLink.writing is set to true, but before SenderLink.Session.SendDelivery() is called in SenderLink.WriteDelivery().

Here is the client code I used to reproduce the issue.

```
class Program
    {
        static void Main(string[] args)
        {
            var address = "amqp://guest:guest@127.0.0.1:5672";
            var conn = new Connection(new Address(address));
            var sess = new Session(conn);
            var link = new SenderLink(sess, "test-client2", "msg-queue");
            int i = 0;

            Console.WriteLine("Press enter to send a message");

            while (true)
            {
                try
                {
                    if (Console.ReadKey().Key == ConsoleKey.Enter)
                    {
                        Console.WriteLine("Sending " + ++i);
                        link.Send(new Message(i % 100));
                    }
                }
                catch (AmqpException)
                {
                    Console.WriteLine("AmqpException thrown");
                }
            }
        }
    }
```
	
Here is the server code I used to reproduce the issue.

```
 class Program
    {
        static void Main(string[] args)
        {
            var address = new Uri("amqp://guest:guest@127.0.0.1:5672");
            ContainerHost host = new ContainerHost(new List<Uri>() { address }, null, address.UserInfo);
            host.RegisterMessageProcessor("msg-queue", new MsgProcessor());
            host.Open();

            while (true)
            {
                Console.WriteLine("Press enter to stop server");

                if (Console.ReadKey().Key == ConsoleKey.Enter)
                {

                    Console.WriteLine("Stopping server");

                    host.Close();
                    host.UnregisterMessageProcessor("msg-queue");

                    Console.WriteLine("Server stopped");
                }

                Console.WriteLine("Press enter to start server");

                if (Console.ReadKey().Key == ConsoleKey.Enter)
                {
                    Console.WriteLine("Starting server");

                    host = new ContainerHost(new List<Uri>() { address }, null, address.UserInfo);
                    host.RegisterMessageProcessor("msg-queue", new MsgProcessor());
                    host.Open();

                    Console.WriteLine("Server started");
                }
            }
        }
    }

    class MsgProcessor : IMessageProcessor
    {
        public int Credit
        {
            get { return 1; }
        }

        public void Process(MessageContext messageContext)
        {
            Thread.Sleep(20);
            Console.WriteLine(messageContext.Message.Body.ToString());
            messageContext.Complete();
        }
    }
```
		
The steps to reproduce are as follows:

1.  Start the server
2.  Start the client
3.  Set a break point in the client in SenderLink.cs at:
		this.Session.SendDeliver(delivery);
	This call is at line 291 in the current code, in the WriteDelivery() method
4.  Press enter in the client console to start sending a message and hit the break point
5.  Press enter in the server console to close the server ContainerHost and unregister the MessageProcessor
6.  Continue execution of the client code
7.  Break all and observe that the pump thread is endlessly spinning in SenderLink.OnAbort() because SenderLink.writing is still set to true

Hopefully this reproduces the problem on your system, and if not, hopefully the description above is enough to see the problem.  I also have some screenshots that I could provide if required.

If there is some way to workaround this issue, let me know.  Thank you for looking into this!

**Comment by xinchen10** [comment]
@mcmellawatt thanks for the detailed bug report. master branch should have a fix that addresses this issue.

**Comment by mcmellawatt** [comment]
My pleasure, thanks for the quick fix.  Any plans for a nuget package release?  It's not urgent, just curious.

**Comment by mcmellawatt** [comment]
Hi Xin,

I don't want to be a bother, but I just wanted to check to see when the next release is planned with this fix.  We are seeing the problem fairly frequently in some of our functional tests.

Thanks!
Ryan

**Comment by xinchen10** [comment]
Fixed in v1.1.7
